compiler/rustc_middle/src/middle/privacy.rs

```diff
@@ -27,26 +28,201 @@ pub enum AccessLevel {
+        EffectiveVisibility { vis: None }
+    }
+
+    pub fn from(visibility: Visibility) -> EffectiveVisibility {
+        EffectiveVisibility { vis: Some(visibility) }
+    }
+
+    pub fn is_public(&self) -> bool {
+        self.vis == Some(Visibility::Public)
+    }
+
+    pub fn is_some(&self) -> bool {
+        match self.vis {
+            Some(_) => true,
+            None => false,
+        }
+    }
+
+    pub fn is_none(&self) -> bool {
+        !self.is_some()
+    }
+
+    pub fn expect_vis(&self) -> Visibility {
+        self.vis.unwrap()
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Debug, HashStable)]
+pub struct EffectiveVisibilityValue {
+    public: EffectiveVisibility,
+    exported: EffectiveVisibility,
+    reachable: EffectiveVisibility,
+    reachable_from_impl_trait: EffectiveVisibility,
+}
+
+impl EffectiveVisibilityValue {
+    pub fn new() -> EffectiveVisibilityValue {
+        EffectiveVisibilityValue {
+            public: EffectiveVisibility::new(),
+            exported: EffectiveVisibility::new(),
+            reachable: EffectiveVisibility::new(),
+            reachable_from_impl_trait: EffectiveVisibility::new(),
+        }
+    }
+
+    pub fn crate_vis() -> EffectiveVisibilityValue {
+        let vis = Visibility::Public;
+        EffectiveVisibilityValue {
+            public: EffectiveVisibility::from(vis),
+            exported: EffectiveVisibility::from(vis),
+            reachable: EffectiveVisibility::from(vis),
+            reachable_from_impl_trait: EffectiveVisibility::from(vis),
+        }
+    }
+
+    pub fn get(&self, tag: AccessLevel) -> &EffectiveVisibility {
+        match tag {
+            AccessLevel::Public => &self.public,
+            AccessLevel::Exported => &self.exported,
+            AccessLevel::Reachable => &self.reachable,
+            AccessLevel::ReachableFromImplTrait => &self.reachable_from_impl_trait,
+        }
+    }
+
+    pub fn get_mut(&mut self, tag: AccessLevel) -> &mut EffectiveVisibility {
+        match tag {
+            AccessLevel::Public => &mut self.public,
+            AccessLevel::Exported => &mut self.exported,
+            AccessLevel::Reachable => &mut self.reachable,
+            AccessLevel::ReachableFromImplTrait => &mut self.reachable_from_impl_trait,
+        }
+    }
+
+    fn merge<T: DefIdTree>(&mut self, vis: EffectiveVisibility, tag: AccessLevel, tree: T) {
+        let current_effective_vis = self.get_mut(tag);
+        if current_effective_vis.is_public() {
+            return;
+        }
+
+        if vis.is_public() || (vis.is_some() && current_effective_vis.is_none()) {
+            current_effective_vis.vis = vis.vis;
+            return;
+        }
+
+        if  let Some(Visibility::Restricted(restricted_id)) = vis.vis &&
+            let Some(Visibility::Restricted(current_restricted_id)) = current_effective_vis.vis {
+            if tree.is_descendant_of(current_restricted_id.to_def_id(), restricted_id.to_def_id()) {
```

Comment:
I think you can reuse `Visibility::is_at_least` instead of going one level deeper and doing this manually.